src/fileparse.py

```diff
@@ -0,0 +1,66 @@
+import csv
+import logging
+from pathlib import Path
+from typing import Any, List, TextIO, Tuple, Union, Callable, Dict, cast
+
+
+def parse_csv(
+    file_path_or_buffer: Union[str, Path, TextIO],
+    select: List[str] = None,
+    types: List[Callable] = None,
+    has_headers: bool = True,
+    delimiter: str = ",",
+    silence_errors=False,
+) -> List[Any]:
```

Comment:
Here you can be little be more precise:
- if `has_headers` is set to `False`, a `List[tuple]` is returned
- of `has_headers` is set to `True`, a `List[Dict[str, Any]]` is returned

So you can replace the static typing of the output by:
```suggestion
) -> Union[List[tuple], List[Dict[str, Any]]]:
```

Reply:
I tried this originally but the `Union` causes issues with containers. For example, I get both these errors when using `Union[List[tuple], List[Dict[str, Any]]]`:
```
error: Argument 1 to "append" of "list" has incompatible type "Dict[str, str]"; expected "Tuple[Any, ...]"
error: Argument 1 to "append" of "list" has incompatible type "Tuple[str, ...]"; expected "Dict[str, Any]"
```

Reply:
Which line is throwing those errors?

Reply:
In `fileparse.py` lines 62 and 64.

Reply:
If you try List[Union[tuple, Dict[str, Any]], do you still have the errors?

Reply:
It fixes the errors in `parse_csv` but it causes issues elsewhere. I get a few of these errors (like in line 14 of `pcost.py`):
```
Incompatible types in assignment (expression has type "List[Union[Tuple[Any, ...], Dict[str, Any]]]", variable has type "List[Dict[str, Any]]")
```
Would I need to change other variables to be less specific like 
`portfolio: List[Dict[str, Any]]` => `portfolio: List[Union[tuple, Dict[str, Any]]]` ?

Reply:
Let's keep `List[Any]` for the output type, it will be easier!

Sometimes mypy is really painful

Reply:
I want to make sure I understand the problem: if I change the return type annotation to `Union[List[tuple], List[Dict[str, Any]]]` then there are these related errors:
```bash
src/report.py:30: error: Argument 1 to "to_portdict" has incompatible type "Union[Tuple[Any, ...], Dict[str, Any]]"; expected "Dict[str, Any]"
src/report.py:42: error: Argument 1 to "dict" has incompatible type "Union[List[Tuple[Any, ...]], List[Dict[str, Any]]]"; expected "Iterable[Tuple[str, float]]"
```

Is this the issue @l-mizrahi?

Reply:
After refactoring `parse_csv` that actually seems to be the only issues. It was worse before.

Reply:
I changed the output type and I managed to fix the typing errors. Needed to `cast` the output of `parse_csv` explicitly in `report.py` and `pcost.py`.

Reply:
Could you elaborate here as well on what you mean by "`cast` the output"?

Reply:
`typing` has a function `cast(<type>, <variable>)` which lets you specify the type of a variable. So from what I understand, if `var` is type `Union[A, B]`, then you can cast it to a specific type (`var=cast(A, var)`) which is then used as its type later on. It seems to solve a lot of the `Union` conflicts.

Reply:
I see, so you are doing `cast(records, List[Dict[str, Any])` in case of `has_headers` set to `True` and `cast(records, List[tuple])` otherwise?

Reply:
Couldn't do it dynamically in `parse_csv`. I don't think mypy can determine types dynamically like that. But, for example, in pcost.py I just used 
```
portfolio = parse_csv(...)
portfolio = cast(List[Dict[str, Any]], portfolio)
```

Reply:
Yeah I think that is fine in this case, I am not sure if an alternative could work like [function overloading](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) or [generics](https://mypy.readthedocs.io/en/stable/generics.html). Otherwise you could also add a type ignore comment: `portdicts = [to_portdict(p) for p in portfolio]  # type: ignore`